pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+
+		#[pallet::constant]
+		type MinimumIncrease: Get<Self::Balance>;
+
+		#[pallet::constant]
+		type AuctionLength: Get<BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type ExtendedLength: Get<BlockNumberFor<Self>>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The specified Auction ID does not exist.
+		InvalidAuctionId,
+		/// The bid price is smaller than the current bid price.
+		BidPriceTooLow,
+		/// When the current auction price exceeds the reserve price, the auction can not be
+		/// canceled.
+		AuctionCancelFailed,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Emitted when an auction has been created and is open for bidding.
+		AuctionCreated { who: T::AccountId, auction_id: AuctionId },
+
+		/// Emitted when an auction has been canceled by its owner.
+		AuctionCanceled { auction_id: AuctionId },
+
+		/// Emitted when an auction has expired without meeting the reserve price.
+		AuctionExpired { auction_id: AuctionId },
+
+		/// Emitted when an auction has concluded successfully with a winning bid.
+		AuctionSucceeded {
+			auction_id: AuctionId,
+			to: T::AccountId,
+			asset: NftId,
+			price: T::Balance,
+		},
+
+		/// Emitted when a new bid is placed on an auction.
+		Bid { who: T::AccountId, auction_id: AuctionId, price: T::Balance },
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_auction)]
+	pub type NextAuctionId<T: Config> = StorageValue<_, AuctionId, ValueQuery>;
+
+	/// Stores the data associated with each auction.
+	#[pallet::storage]
+	#[pallet::getter(fn auctions)]
+	pub type Auctions<T: Config> = StorageMap<_, Blake2_128Concat, AuctionId, AuctionData<T>>;
+
+	/// Stores the auction IDs that are set to expire at a specific block.
+	#[pallet::storage]
+	pub type AuctionsExpiryBlock<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, Vec<AuctionId>, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(block_number: BlockNumberFor<T>) {
+			// Expire auctions that are due.
+			AuctionsExpiryBlock::<T>::take(block_number).into_iter().for_each(|auction_id| {
+				if let Some(auction_data) = Auctions::<T>::take(auction_id) {
+					Self::resolve_auction(auction_id, auction_data);
+				}
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates a new auction for a specified NFT.
+		/// This function can only be called by an account with the Customer role.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::create_auction())]
+		pub fn create_auction(
+			origin: OriginFor<T>,
+			nft_id: NftId,
+			start: Option<T::Balance>,
+			reserve: Option<T::Balance>,
+			buy_now: Option<T::Balance>,
+		) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
```

Comment:
The role check is not required. Only customer is allowed to own NFT. 
We need to allow managers to auction NFT on behalf of the bank. This can be done in a later ticket.